// cl2-component-library/src/components/Select/index.tsx
import React, { PureComponent, FocusEvent, ChangeEvent } from 'react';
import { isString, get } from 'lodash';
import Icon from '../Icon';
import Label from '../Label';
import IconTooltip from '../IconTooltip';
import { IOption, InputSize } from '../../utils/typings';
import styled from 'styled-components';
import { defaultInputStyle, colors, isRtl } from '../../utils/styleUtils';
import testEnv from '../../utils/testUtils/testEnv';

const Arrow = styled(Icon)`
  width: 12px;
  height: 12px;
  fill: #999;
  pointer-events: none;
  transform: rotate(90deg);
  margin: auto;
  position: absolute;
  top: 0;
  bottom: 0;
  right: 11px;

  ${isRtl`
    right: auto;
    left: 11px;
  `}
`;

const Container = styled.div`
  position: relative;
  outline: none !important;

  &.enabled {
    &:hover {
      ${Arrow} {
        fill: ${colors.hoveredBorder};
      }
    }

    &:focus {
      ${Arrow} {
        fill: ${colors.focussedBorder};
      }
    }
  }

  &.disabled {
    ${Arrow} {
      fill: #666;
    }
  }
`;

const SelectWrapper = styled.div`
  width: 100%;
  position: relative;
  select {
    width: 100%;
    margin: 0;
    padding-right: 27px;
    cursor: pointer;
    outline: none !important;

    ${isRtl`
    padding-right: 0;
    padding-left: 27px;
  `}
    &::-ms-expand {
      display: none;
    }
    ${defaultInputStyle};
  }
`;

export interface DefaultProps {
  canBeEmpty?: boolean;
}

export interface Props extends DefaultProps {
  id?: string;
  disabled?: boolean;
  onChange: (arg: IOption) => void;
  onBlur?: (event: FocusEvent<HTMLSelectElement>) => void;
  value?: IOption | string | null;
  options: IOption[] | null;
  label?: string | JSX.Element | null;
  labelTooltipText?: string | JSX.Element | null;
  className?: string;
  size?: InputSize;
}

class Select extends PureComponent<Props> {
  static defaultProps: DefaultProps = {
    canBeEmpty: false,
  };

  handleOnChange = (event: ChangeEvent<HTMLSelectElement>) => {
    if (this.props.options) {
      const selectedOption = this.props.options.find(
        (option) => option.value.toString() === event.target.value.toString()
      ) as IOption;
      this.props.onChange(selectedOption);
    }
  };

  handleOnBlur = (event: FocusEvent<HTMLSelectElement>) => {
    if (this.props.onBlur) {
      this.props.onBlur(event);
    }
  };

  render() {
    const {
      id,
      disabled,
      className,
      options,
      canBeEmpty,
      label,
      labelTooltipText,
      size,
    } = this.props;
    const defaultValue = 'DEFAULT_SELECT_VALUE';
    const value = isString(this.props.value)
      ? this.props.value
      : (get(this.props.value, 'value', null) as string | null);
    const selectedValue =
      options && options.find((option) => option.value === value)
        ? (options.find((option) => option.value === value)?.value as string)
        : defaultValue;

    return (
      <Container
        className={`${className} ${disabled ? 'disabled' : 'enabled'}`}
      >
        {label && (
          <Label htmlFor={id}>
            <span>{label}</span>
            {labelTooltipText && <IconTooltip content={labelTooltipText} />}
          </Label>
        )}
        <SelectWrapper size={size}>
          <select
            id={id}
            disabled={disabled}
            onChange={this.handleOnChange}
            onBlur={this.handleOnBlur}
            className={disabled ? 'disabled' : 'enabled'}
            value={selectedValue}
            data-testid={testEnv('select')}
          >
            <option
              value={defaultValue}
              aria-selected={selectedValue === defaultValue}
              hidden={!canBeEmpty}
              disabled={!canBeEmpty}
            />

            {options &&
              options.length > 0 &&
              options.map((option, index) => (
                <option
                  key={index}
                  value={option.value}
                  aria-selected={selectedValue === option.value}
                >
                  {option.label}
                </option>
              ))}
          </select>
          <Arrow
            name="chevron-right"
            ariaHidden
            className={`arrow ${disabled ? 'disabled' : 'enabled'}`}
          />
        </SelectWrapper>
      </Container>
    );
  }
}

export default Select;
