// cl2-component-library/src/components/LocaleSwitcher/index.tsx
import React, { PureComponent, MouseEvent } from 'react';
import { isEmpty, get } from 'lodash';
import styled from 'styled-components';
import { rgba } from 'polished';
import { colors, fontSizes, isRtl } from '../../utils/styleUtils';
import { MultilocFormValues, Locale } from '../../utils/typings';

const Container = styled.div`
  width: 100%;
  display: flex;
  justify-content: flex-end;
`;

const StyledButton = styled.button`
  color: ${colors.adminTextColor};
  font-size: ${fontSizes.small}px;
  display: flex;
  align-items: center;
  text-transform: uppercase;
  font-weight: 500;
  white-space: nowrap;
  padding: 7px 8px;
  margin-right: 6px;
  border-radius: ${(props: any) => props.theme.borderRadius};
  background: ${colors.lightGreyishBlue};
  cursor: pointer;
  transition: all 80ms ease-out;

  ${isRtl`
    flex-direction: row-reverse;
  `}

  &.last {
    margin-right: 0px;
  }

  &:not(.selected):hover {
    color: ${colors.adminTextColor};
    background: ${rgba(colors.adminTextColor, 0.2)};
  }

  &.selected {
    color: #fff;
    background: ${colors.adminTextColor};
  }
`;

const Dot = styled.div`
  flex: 0 0 9px;
  width: 9px;
  height: 9px;
  border-radius: 50%;
  background: ${colors.clRed};
  margin-right: 5px;

  ${isRtl`
    margin-right: 0px;
    margin-left: 5px;
  `}

  &.notEmpty {
    background: ${colors.clGreen};
  }
`;

interface Props {
  onSelectedLocaleChange: (selectedLocale: Locale) => void;
  locales: Locale[];
  selectedLocale: Locale;
  values?: MultilocFormValues;
  className?: string;
}

class LocaleSwitcher extends PureComponent<Props> {
  isValueForLocaleNotEmpty = (locale: Locale) => {
    const { values } = this.props;
    return Object.getOwnPropertyNames(values).every(
      (key) => !isEmpty(get(values, `[${key}][${locale}]`))
    );
  };

  removeFocus = (event: MouseEvent<HTMLButtonElement>) => {
    event.preventDefault();
  };

  handleOnClick = (locale: Locale) => (
    event: MouseEvent<HTMLButtonElement>
  ) => {
    event.preventDefault();
    if (this.props.selectedLocale !== locale) {
      this.props.onSelectedLocaleChange(locale);
    }
  };

  render() {
    const { locales, selectedLocale, values, className } = this.props;

    if (locales && locales.length > 1) {
      return (
        <Container className={className}>
          {locales.map((locale, index) => (
            <StyledButton
              key={locale}
              onMouseDown={this.removeFocus}
              onClick={this.handleOnClick(locale)}
              type="button"
              className={[
                'e2e-localeswitcher',
                locale,
                locale === selectedLocale ? 'selected' : '',
                index + 1 === locales.length ? 'last' : '',
              ].join(' ')}
            >
              {values && (
                <Dot
                  className={
                    this.isValueForLocaleNotEmpty(locale) ? 'notEmpty' : 'empty'
                  }
                />
              )}
              {locale}
            </StyledButton>
          ))}
        </Container>
      );
    }

    return null;
  }
}

export default LocaleSwitcher;
