// cl2-component-library/src/components/Table/index.tsx
import React, { PureComponent } from 'react';
import styled from 'styled-components';
import { darken } from 'polished';
import { fontSizes, colors } from '../../utils/styleUtils';

const StyledTable = styled.table`
  width: 100%;
  padding: 3px;
  margin: 0;
  border: none;
  border-spacing: 0;
  border-collapse: collapse;
  table-layout: auto;

  th,
  td {
    padding: 0;
    margin: 0;
    text-align: left;
  }

  thead {
    tr {
      th {
        fill: ${colors.text};
        color: ${colors.adminTextColor};
        font-size: ${fontSizes.small}px;
        font-weight: 400;
        line-height: normal;
        padding-top: 0px;
        padding-bottom: 15px;
        border-bottom: solid 1px #ccc;

        div {
          display: flex;
          align-items: flex-start;
        }
      }
    }
  }

  tbody {
    tr {
      border-radius: ${(props: any) => props.theme.borderRadius};
      border-bottom: solid 1px #ddd;
      td {
        color: ${colors.adminTextColor};
        font-size: ${fontSizes.base}px;
        font-weight: 500;
        line-height: normal;
        text-align: left;
        padding-top: 5px;
        padding-bottom: 5px;
        &.center {
          justify-content: center;
        }

        a {
          color: ${colors.clBlue};
          text-decoration: underline;

          &:hover {
            color: ${darken(0.2, colors.clBlue)};
            text-decoration: underline;
          }
        }
        & > * {
          margin-right: 5px;
        }
      }
    }
  }

  tfoot {
    display: flex;

    tr {
      display: flex;

      td {
        display: flex;
      }
    }
  }
`;

interface Props {
  className?: string;
}

export default class Table extends PureComponent<Props> {
  render() {
    const { className, children } = this.props;

    return (
      <StyledTable cellSpacing="0" cellPadding="0" className={className}>
        {children}
      </StyledTable>
    );
  }
}
