// cl2-component-library/src/components/StatusLabel/index.tsx
import React, { FC, memo } from 'react';
import styled from 'styled-components';
import { fontSizes } from '../../utils/styleUtils';
import Icon, { Props as IconProps } from '../Icon';

const Container = styled.div<{ backgroundColor: string }>`
  height: 28px;
  color: #fff;
  font-size: ${fontSizes.xs}px;
  font-weight: 500;
  line-height: normal;
  text-transform: uppercase;
  white-space: nowrap;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  border-radius: ${(props: any) => props.theme.borderRadius};
  background: ${(props) => props.backgroundColor};
  padding-left: 10px;
  padding-right: 10px;
`;

const StyledIcon = styled(Icon)`
  height: 13px;
  margin-right: 8px;
  fill: #fff;
`;

interface Props {
  className?: string;
  text: JSX.Element | string;
  backgroundColor: string;
  icon?: IconProps['name'];
}

const StatusLabel: FC<Props> = memo<Props>(
  ({ backgroundColor, className, icon, text }) => {
    return (
      <Container backgroundColor={backgroundColor} className={className || ''}>
        {icon && <StyledIcon name={icon} />}
        {text}
      </Container>
    );
  }
);

export default StatusLabel;
