// cl2-component-library/src/components/Button/index.tsx
import React, { PureComponent, MouseEvent, ButtonHTMLAttributes } from 'react';
import { isBoolean, isNil, isString, get } from 'lodash';
import styled from 'styled-components';
import { darken, lighten, transparentize, opacify, rgba } from 'polished';
import {
  colors,
  invisibleA11yText,
  fontSizes,
  defaultStyles,
  isRtl,
} from '../../utils/styleUtils';
import Spinner from '../Spinner';
import Link from '../Link';
import Icon, { Props as IconProps } from '../Icon';
import { Locale } from '../../utils/typings';
import Box, {
  BoxMarginProps,
  BoxPaddingProps,
  BoxPositionProps,
  BoxWidthProps,
  BoxHeightProps,
} from '../Box';

export type ButtonStyles =
  | 'primary'
  | 'primary-inverse'
  | 'primary-outlined'
  | 'secondary'
  | 'secondary-outlined'
  | 'white'
  | 'success'
  | 'text'
  | 'cl-blue'
  | 'cl-blue-outlined'
  | 'admin-dark'
  | 'admin-dark-outlined'
  | 'admin-dark-text'
  | 'delete';

type DefaultStyleValues = {
  [key in ButtonStyles]: {
    bgColor: string;
    bgHoverColor?: string;
    textColor: string;
    textHoverColor?: string;
    borderColor?: string;
    borderHoverColor?: string;
    boxShadow?: string;
    boxShadowHover?: string;
    iconColor?: string;
    iconHoverColor?: string;
    padding?: string;
  };
};

function getFontSize(props: ButtonContainerProps & { theme: any }) {
  if (props.fontSize) {
    return props.fontSize;
  } else {
    switch (props.size) {
      case '2':
        return `${fontSizes.large}px`;
      case '3':
        return `${fontSizes.xl}px`;
      default:
        return `${fontSizes.base}px`;
    }
  }
}

function getPadding(props: ButtonContainerProps & { theme: any }) {
  if (props.padding) {
    return props.padding;
  } else if (props.p) {
    return props.p;
  } else {
    switch (props.size) {
      case '2':
        return '11px 22px';
      case '3':
        return '13px 24px';
      case '4':
        return '15px 26px';
      default:
        // return '.65em 1.45em';
        return '9px 18px';
    }
  }
}

function getIconHeight(props: ButtonContainerProps & { theme: any }) {
  if (props.iconSize) {
    return props.iconSize;
  } else {
    switch (props.size) {
      case '2':
        return '18px';
      case '3':
        return '19px';
      case '4':
        return '20px';
      default:
        return '17px';
    }
  }
}

function getLineHeight(props: ButtonContainerProps & { theme: any }) {
  if (props.lineHeight) {
    return props.lineHeight;
  } else {
    switch (props.size) {
      case '2':
        return '24px';
      case '3':
        return '26px';
      case '4':
        return '28px';
      default:
        return '22px';
    }
  }
}

function getButtonStyle(props: ButtonContainerProps & { theme: any }) {
  const defaultStyleValues: DefaultStyleValues = {
    primary: {
      bgColor: get(props.theme, 'colorMain'),
      textColor: '#fff',
      textHoverColor: '#fff',
      iconColor: '#fff',
      iconHoverColor: '#fff',
    },
    'primary-outlined': {
      bgColor: 'transparent',
      bgHoverColor: transparentize(0.95, get(props.theme, 'colorMain')),
      textColor: get(props.theme, 'colorMain'),
      borderColor: get(props.theme, 'colorMain'),
    },
    'primary-inverse': {
      bgColor: '#fff',
      textColor: get(props.theme, 'colorText'),
      textHoverColor: get(props.theme, 'colorText'),
    },
    secondary: {
      bgColor: colors.lightGreyishBlue,
      textColor: darken(0.1, colors.label),
      bgHoverColor: darken(0.05, colors.lightGreyishBlue),
    },
    'secondary-outlined': {
      bgColor: 'transparent',
      bgHoverColor: transparentize(0.95, colors.label),
      textColor: colors.label,
      borderColor: lighten(0.25, colors.label),
    },
    white: {
      bgColor: '#fff',
      bgHoverColor: '#fff',
      iconColor: colors.label,
      textColor: get(props.theme, 'colorText'),
      borderColor: 'transparent',
      boxShadow: defaultStyles.boxShadow,
      boxShadowHover: defaultStyles.boxShadowHoverSmall,
    },
    text: {
      bgColor: 'transparent',
      bgHoverColor: 'transparent',
      textColor: colors.label,
      iconColor: colors.label,
    },
    success: {
      bgColor: colors.clGreenSuccessBackground,
      textColor: colors.clGreenSuccess,
    },
    'cl-blue': {
      bgColor: colors.clBlueDark,
      textColor: '#fff',
      textHoverColor: '#fff',
      iconColor: '#fff',
      iconHoverColor: '#fff',
    },
    'cl-blue-outlined': {
      bgColor: 'transparent',
      bgHoverColor: 'transparent',
      textColor: colors.clBlueDark,
      borderColor: colors.clBlueDark,
    },
    'admin-dark': {
      bgColor: colors.adminTextColor,
      textColor: '#fff',
      textHoverColor: '#fff',
      iconColor: '#fff',
      iconHoverColor: '#fff',
    },
    'admin-dark-outlined': {
      bgColor: 'transparent',
      bgHoverColor: 'transparent',
      textColor: colors.adminTextColor,
      borderColor: colors.adminTextColor,
    },
    'admin-dark-text': {
      bgColor: 'transparent',
      bgHoverColor: rgba(colors.adminTextColor, 0.1),
      textColor: colors.adminTextColor,
    },
    delete: {
      bgColor: colors.clRedError,
      textColor: '#fff',
      textHoverColor: '#fff',
      iconColor: '#fff',
      iconHoverColor: '#fff',
    },
  };

  const backgroundColor =
    props.bgColor || get(defaultStyleValues, `${props.buttonStyle}.bgColor`);

  const backgroundHoverColor =
    props.bgHoverColor ||
    get(defaultStyleValues, `${props.buttonStyle}.bgHoverColor`) ||
    darken(0.12, backgroundColor);
  const backgroundActiveColor =
    backgroundHoverColor !== 'transparent'
      ? backgroundHoverColor?.startsWith('rgba')
        ? opacify(0.12, backgroundHoverColor)
        : darken(0.12, backgroundHoverColor)
      : 'inherit';
  const backgroundDisabledColor =
    props.bgDisabledColor ||
    get(defaultStyleValues, `${props.buttonStyle}.bgDisabledColor`) ||
    backgroundColor;
  const textColor =
    props.textColor ||
    get(defaultStyleValues, `${props.buttonStyle}.textColor`);
  const textHoverColor =
    props.textHoverColor ||
    get(defaultStyleValues, `${props.buttonStyle}.textHoverColor`) ||
    darken(0.2, textColor);
  const iconColor =
    props.iconColor ||
    get(defaultStyleValues, `${props.buttonStyle}.iconColor`) ||
    textColor;
  const iconHoverColor =
    props.iconHoverColor ||
    get(defaultStyleValues, `${props.buttonStyle}.iconHoverColor`) ||
    darken(0.2, iconColor);
  const textDisabledColor =
    props.textDisabledColor ||
    get(defaultStyleValues, `${props.buttonStyle}.textDisabledColor`) ||
    textColor;
  const borderColor =
    props.borderColor ||
    get(defaultStyleValues, `${props.buttonStyle}.borderColor`) ||
    'transparent';
  const borderHoverColor =
    props.borderHoverColor ||
    get(defaultStyleValues, `${props.buttonStyle}.borderHoverColor`) ||
    darken(0.2, borderColor);
  const borderDisabledColor =
    props.borderDisabledColor ||
    get(defaultStyleValues, `${props.buttonStyle}.borderDisabledColor`) ||
    borderColor;
  const boxShadow =
    props.boxShadow ||
    get(defaultStyleValues, `${props.buttonStyle}.boxShadow`) ||
    'none';
  const boxShadowHover =
    props.boxShadowHover ||
    get(defaultStyleValues, `${props.buttonStyle}.boxShadowHover`) ||
    'none';
  const borderRadius =
    props.borderRadius ||
    get(defaultStyleValues, `${props.buttonStyle}.borderRadius`) ||
    props.theme.borderRadius;
  const textDecoration =
    props.textDecoration ||
    get(defaultStyleValues, `${props.buttonStyle}.textDecoration`) ||
    'none';
  const textDecorationHover =
    props.textDecorationHover ||
    get(defaultStyleValues, `${props.buttonStyle}.textDecorationHover`) ||
    'none';
  const padding =
    get(defaultStyleValues, `${props.buttonStyle}.padding`) ||
    getPadding(props);
  const fontSize = getFontSize(props);
  const lineHeight = getLineHeight(props);
  const iconSize = getIconHeight(props);
  const fontWeight = props.fontWeight || 'normal';
  const borderWidth = props.borderThickness || '1px';
  const display = !props.width ? 'inline-flex' : 'flex';
  const height = props.height || props.h || 'auto';
  const minHeight = props.minHeight || 'initial';
  const maxHeight = props.maxHeight || 'initial';
  const justifyContent = props.justify || 'center';
  const minWidth = props.minWidth || 'initial';
  const maxWidth = props.maxWidth || 'initial';
  const width = props.width || props.w || '100%';
  const buttonTextOpacity = props.processing ? 0 : 1;
  const iconOpacity = props.processing ? 0 : 1;
  const whiteSpace = props.whiteSpace || 'nowrap';
  const opacityDisabled = props.opacityDisabled || '0.37';
  const flexDirection = props.theme.isRtl ? 'row-reverse' : 'row';

  return `
    width: ${width};
    min-width: ${minWidth};
    max-width: ${maxWidth};
    height: ${height};
    min-height: ${minHeight};
    max-height: ${maxHeight};
    display: ${display};
    align-items: center;
    justify-content: ${justifyContent};
    margin: 0;
    position: relative;
    border-radius: ${borderRadius};
    background: ${backgroundColor};
    border-width: ${borderWidth};
    border-style: solid;
    border-color: ${borderColor};
    box-shadow: ${boxShadow};
    cursor: pointer;
    transition: background 80ms ease-out, border-color 80ms ease-out, box-shadow 125ms ease-in-out 0s;
    flex-direction: ${flexDirection};

    // padding
    padding: ${padding};
    ${props.paddingY ? `padding-top: ${props.paddingY}` : ''};
    ${props.py ? `padding-top: ${props.py}` : ''};
    ${props.paddingTop ? `padding-top: ${props.paddingTop}` : ''};
    ${props.pt ? `padding-top: ${props.pt}` : ''};
    ${props.paddingY ? `padding-bottom: ${props.paddingY}` : ''};
    ${props.py ? `padding-bottom: ${props.py}` : ''};
    ${props.paddingBottom ? `padding-bottom: ${props.paddingBottom}` : ''};
    ${props.pb ? `padding-bottom: ${props.pb}` : ''};
    ${props.paddingX ? `padding-left: ${props.paddingX}` : ''};
    ${props.px ? `padding-left: ${props.px}` : ''};
    ${props.paddingLeft ? `padding-left: ${props.paddingLeft}` : ''};
    ${props.pl ? `padding-left: ${props.pl}` : ''};
    ${props.paddingX ? `padding-right: ${props.paddingX}` : ''};
    ${props.px ? `padding-right: ${props.px}` : ''};
    ${props.paddingRight ? `padding-right: ${props.paddingRight}` : ''};
    ${props.pr ? `padding-right: ${props.pr}` : ''};

    ${ButtonText} {
      color: ${textColor};
      opacity: ${buttonTextOpacity};
      font-size: ${fontSize};
      line-height: ${lineHeight};
      font-weight: ${fontWeight};
      text-decoration: ${textDecoration};
      white-space: ${whiteSpace};
      text-align: left;
      margin: 0;
      margin-top: -1px;
      padding: 0;
      transition: color 80ms ease-out;

      ${isRtl`
        text-align: right;
      `}
    }

    ${StyledIcon} {
      fill: ${iconColor};
      flex: 0 0 ${iconSize};
      height: ${iconSize};
      width: ${iconSize};
      opacity: ${iconOpacity};
      transition: fill 80ms ease-out;
    }

    &:not(.disabled):not(.processing):active {
      background: ${backgroundActiveColor} !important;
    }

    &:not(.disabled):not(.processing):hover {
      background: ${backgroundHoverColor};
      border-color: ${borderHoverColor};
      box-shadow: ${boxShadowHover};

      ${ButtonText} {
        color: ${textHoverColor};
        text-decoration: ${textDecorationHover};
      }

      ${StyledIcon} {
        fill: ${iconHoverColor};
      }
    }

    &.fullWidth {
      flex: 1;
      width: 100%;
    }

    &.disabled {
      background: ${backgroundDisabledColor};
      border-color: ${borderDisabledColor};
      opacity: ${opacityDisabled};
      cursor: not-allowed;

      ${ButtonText} {
        color: ${textDisabledColor};
      }

      ${StyledIcon} {
        fill: ${textDisabledColor};
      }
    }

    & .spinner {
      border-color: ${textColor} !important;
      border-right-color: transparent !important;
    }
  `;
}

const StyledButton = styled.button``;
const StyledLink = styled(Link)``;
const StyledA = styled.a``;
const ButtonText = styled.div``;
const StyledIcon = styled(Icon)`
  &.hasText {
    &.left {
      margin-right: 10px;
    }

    &.right {
      margin-left: 10px;
    }
  }

  ${isRtl`
        &.hasText {
            &.left {
            margin-right: 0px;
            margin-left: 10px;
            }

            &.right {
            margin-left: 0px;
            margin-right: 10px;
            }
        }
        `}
`;

const Container = styled(Box)<ButtonContainerProps>`
  display: flex;
  align-items: center;
  justify-content: ${(props) => props.justifyWrapper || 'center'};
  padding: 0;
  user-select: none;
  background-color: transparent;

  &.fullWidth {
    width: 100%;
  }

  ${StyledButton},
  ${StyledLink},
  ${StyledA} {
    ${(props) => getButtonStyle(props)}
  }
`;

const SpinnerWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0px;
`;

const HiddenText = styled.span`
  ${invisibleA11yText()}
`;

type Size = '1' | '2' | '3' | '4' | undefined;

export type ButtonContainerProps = {
  buttonStyle?: ButtonStyles;
  size?: Size;
  width?: string;
  height?: string;
  fullWidth?: boolean;
  padding?: string;
  justify?: 'left' | 'center' | 'right' | 'space-between';
  justifyWrapper?: 'left' | 'center' | 'right' | 'space-between';
  iconSize?: string;
  processing?: boolean;
  disabled?: boolean;
  iconColor?: string;
  iconHoverColor?: string;
  textColor?: string;
  textHoverColor?: string;
  textDisabledColor?: string;
  bgColor?: string;
  bgHoverColor?: string;
  bgDisabledColor?: string;
  borderColor?: string;
  borderHoverColor?: string;
  borderDisabledColor?: string;
  borderThickness?: string;
  boxShadow?: string;
  boxShadowHover?: string;
  borderRadius?: string;
  fontWeight?: string;
  lineHeight?: string;
  textDecoration?: string;
  textDecorationHover?: string;
  whiteSpace?: string;
  minWidth?: string;
  fontSize?: string;
  opacityDisabled?: string;
  onClick?: (arg: MouseEvent) => void;
} & BoxMarginProps &
  BoxPaddingProps &
  BoxPositionProps &
  BoxWidthProps &
  BoxHeightProps &
  React.HTMLAttributes<HTMLDivElement>;

export interface Props extends ButtonContainerProps {
  children?: any;
  className?: string;
  locale: Locale;
  form?: string;
  hiddenText?: string | JSX.Element;
  icon?: IconProps['name'];
  iconPos?: 'left' | 'right';
  iconTitle?: IconProps['title'];
  linkTo?: string;
  setSubmitButtonRef?: (value: any) => void;
  text?: string | JSX.Element;
  theme?: Record<string, any> | undefined;
  type?: ButtonHTMLAttributes<HTMLButtonElement>['type'];
  spinnerColor?: string;
  role?: string;
  ariaLabel?: string;
  autoFocus?: boolean;
  fontSize?: string;
  ariaExpanded?: boolean;
  ariaDescribedby?: string;
  iconAriaHidden?: boolean;
  ariaDisabled?: boolean;
  openLinkInNewTab?: boolean;
}

class Button extends PureComponent<Props> {
  handleOnClick = (event: any) => {
    const { onClick, processing, disabled } = this.props;

    if (onClick) {
      event.preventDefault();
      event.stopPropagation();

      if (!disabled && !processing) {
        onClick(event);
      }
    }
  };

  removeFocus = (event: MouseEvent<HTMLElement>) => {
    event.preventDefault();
  };

  getSpinnerSize = (size: Size) => {
    switch (size) {
      case '2':
        return '26px';
      case '3':
        return '28px';
      case '4':
        return '30px';
      default:
        return '24px';
    }
  };

  render() {
    const {
      type,
      text,
      form,
      iconColor,
      iconHoverColor,
      textColor,
      textHoverColor,
      textDisabledColor,
      bgColor,
      bgHoverColor,
      bgDisabledColor,
      borderColor,
      borderHoverColor,
      borderDisabledColor,
      borderThickness,
      boxShadow,
      boxShadowHover,
      borderRadius,
      justify,
      justifyWrapper,
      icon,
      iconSize,
      iconTitle,
      hiddenText,
      children,
      linkTo,
      fontWeight,
      lineHeight,
      textDecoration,
      textDecorationHover,
      whiteSpace,
      fullWidth,
      role,
      ariaLabel,
      fontSize,
      autoFocus,
      ariaExpanded,
      ariaDescribedby,
      ariaDisabled,
      opacityDisabled,
      className,
      openLinkInNewTab,
      onClick: _onClick,
      ...rest
    } = this.props;

    const id = this.props.id || '';
    const size = this.props.size || '1';
    const iconAriaHidden = isBoolean(this.props.iconAriaHidden)
      ? this.props.iconAriaHidden
      : true;
    const processing = isBoolean(this.props.processing)
      ? this.props.processing
      : false;
    const buttonStyle = this.props.buttonStyle || 'primary';
    const disabled = isBoolean(this.props.disabled)
      ? this.props.disabled
      : false;
    const iconPos = this.props.iconPos || 'left';
    const spinnerSize = this.getSpinnerSize(size);
    const hasText = !isNil(text) || !isNil(children);
    const containerClassNames = [
      className,
      disabled ? 'disabled' : null,
      processing ? 'processing' : null,
      fullWidth ? 'fullWidth' : null,
    ]
      .filter((item: any) => !isNil(item))
      .join(' ');
    const buttonClassNames = [
      'button',
      'Button',
      buttonStyle,
      disabled ? 'disabled' : null,
      processing ? 'processing' : null,
      fullWidth ? 'fullWidth' : null,
    ]
      .filter((item: any) => !isNil(item))
      .join(' ');

    const childContent = (
      <>
        {icon && iconPos === 'left' && (
          <StyledIcon
            name={icon}
            className={`buttonIcon ${iconPos} ${hasText && 'hasText'}`}
            title={iconTitle}
            ariaHidden={iconAriaHidden}
          />
        )}
        {hasText && (
          <ButtonText className="buttonText">{text || children}</ButtonText>
        )}
        {hiddenText && <HiddenText>{hiddenText}</HiddenText>}
        {icon && iconPos === 'right' && (
          <StyledIcon
            name={icon}
            className={`buttonIcon ${iconPos} ${hasText && 'hasText'}`}
            title={iconTitle}
            ariaHidden={iconAriaHidden}
          />
        )}
        {processing && (
          <SpinnerWrapper>
            <Spinner size={spinnerSize} />
          </SpinnerWrapper>
        )}
      </>
    );

    return (
      <Container
        onClick={this.handleOnClick}
        className={containerClassNames}
        onMouseDown={this.removeFocus}
        buttonStyle={buttonStyle}
        id={id}
        size={size}
        justify={justify}
        justifyWrapper={justifyWrapper}
        iconSize={iconSize}
        processing={processing}
        disabled={disabled}
        iconColor={iconColor}
        iconHoverColor={iconHoverColor}
        textColor={textColor}
        textHoverColor={textHoverColor}
        textDisabledColor={textDisabledColor}
        bgColor={bgColor}
        bgHoverColor={bgHoverColor}
        bgDisabledColor={bgDisabledColor}
        borderColor={borderColor}
        borderHoverColor={borderHoverColor}
        borderDisabledColor={borderDisabledColor}
        borderThickness={borderThickness}
        boxShadow={boxShadow}
        boxShadowHover={boxShadowHover}
        borderRadius={borderRadius}
        fontWeight={fontWeight}
        lineHeight={lineHeight}
        textDecoration={textDecoration}
        textDecorationHover={textDecorationHover}
        whiteSpace={whiteSpace}
        fontSize={fontSize}
        opacityDisabled={opacityDisabled}
        {...rest}
      >
        {linkTo && !disabled ? (
          isString(linkTo) &&
          (linkTo.startsWith('http') || linkTo.startsWith('www')) ? (
            <StyledA
              ref={this.props.setSubmitButtonRef}
              href={linkTo}
              className={buttonClassNames}
              role={role}
              aria-label={ariaLabel}
              target={openLinkInNewTab ? '_blank' : undefined}
              rel={openLinkInNewTab ? 'noreferrer' : undefined}
            >
              {childContent}
            </StyledA>
          ) : (
            <StyledLink
              ref={this.props.setSubmitButtonRef}
              locale={this.props.locale}
              to={linkTo}
              className={buttonClassNames}
              role={role}
              aria-label={ariaLabel}
              target={openLinkInNewTab ? '_blank' : undefined}
            >
              {childContent}
            </StyledLink>
          )
        ) : (
          <StyledButton
            role={role}
            aria-label={ariaLabel}
            aria-expanded={ariaExpanded}
            aria-describedby={ariaDescribedby}
            aria-disabled={ariaDisabled}
            disabled={disabled}
            ref={this.props.setSubmitButtonRef}
            className={buttonClassNames}
            form={form}
            type={type ? type : 'submit'}
            // eslint-disable-next-line jsx-a11y/no-autofocus
            autoFocus={autoFocus}
          >
            {childContent}
          </StyledButton>
        )}
      </Container>
    );
  }
}

export default Button;
